File: src/engine/board.test.ts
import { describe, expect, it } from 'vitest'
import {
  boardHash,
  createGame,
  legalMoves,
  pass,
  pointLabel,
  rebuildPositionHashes,
  replayHistory,
  starPoints,
  tryPlay,
} from './board'
import type { Stone } from './types'

describe('boardHash', () => {
  it('is injective over the states we care about', () => {
    const seen = new Map<string, string>()
    // 9칸 판을 3진 전수 탐색 — 19683가지 배치가 전부 다른 해시여야 한다
    for (let n = 0; n < 3 ** 9; n++) {
      const board: Stone[] = []
      let v = n
      for (let i = 0; i < 9; i++) {
        board.push((v % 3) as Stone)
        v = Math.floor(v / 3)
      }
      const h = boardHash(board)
      expect(seen.has(h)).toBe(false)
      seen.set(h, board.join(''))
    }
    expect(seen.size).toBe(3 ** 9)
  })

  it('stays inside the safe BMP range (no lone surrogates in localStorage)', () => {
    const board = Array<Stone>(19 * 19).fill(2)
    const h = boardHash(board)
    for (const ch of h) {
      const code = ch.charCodeAt(0)
      expect(code).toBeLessThan(0xd800)
      expect(code).toBeGreaterThanOrEqual(0x30)
    }
    // JSON 왕복이 손실 없이 되는지 (스냅샷 저장 경로)
    expect(JSON.parse(JSON.stringify({ h })).h).toBe(h)
  })

  it('is much shorter than the old join()', () => {
    const board = Array<Stone>(19 * 19).fill(1)
    expect(boardHash(board).length).toBeLessThan(board.length / 5)
  })
})

describe('rebuildPositionHashes', () => {
  it('restores superko protection when the stored hashes are garbage', () => {
    let g = createGame(9)
    for (const [x, y] of [[3, 3], [5, 5], [3, 4], [5, 4]] as [number, number][]) {
      const r = tryPlay(g, x, y)
      expect(r.ok).toBe(true)
      if (r.ok) g = r.state
    }
    const corrupted = { ...g, positionHashes: ['nonsense'] }
    const fixed = rebuildPositionHashes(corrupted)
    expect(fixed.positionHashes).toEqual(g.positionHashes)
    // history 밖 정보는 보존
    expect(fixed.ended).toBe(g.ended)
    expect(fixed.board).toEqual(g.board)
  })

  it('replayHistory reproduces the live board', () => {
    let g = createGame(9)
    for (const [x, y] of [[2, 2], [6, 6], [2, 3], [6, 5], [4, 4]] as [number, number][]) {
      const r = tryPlay(g, x, y)
      if (r.ok) g = r.state
    }
    const replayed = replayHistory(9, g.history)
    expect(replayed.board).toEqual(g.board)
    expect(replayed.toPlay).toBe(g.toPlay)
    expect(replayed.positionHashes).toEqual(g.positionHashes)
  })
})

describe('legalMoves', () => {
  it('excludes occupied points and matches tryPlay exactly', () => {
    let g = createGame(9)
    for (const [x, y] of [[4, 4], [3, 3], [5, 5]] as [number, number][]) {
      const r = tryPlay(g, x, y)
      if (r.ok) g = r.state
    }
    const fast = new Set(legalMoves(g).map((p) => `${p.x},${p.y}`))
    for (let y = 0; y < 9; y++) {
      for (let x = 0; x < 9; x++) {
        expect(fast.has(`${x},${y}`)).toBe(tryPlay(g, x, y).ok)
      }
    }
    expect(fast.size).toBe(81 - 3)
  })
})

describe('pointLabel', () => {
  it('puts row 1 at the bottom (standard Go notation)', () => {
    // board 배열은 y=0 이 위쪽 줄 → 세로 반전
    expect(pointLabel(0, 18, 19)).toBe('A1') // 좌하단
    expect(pointLabel(0, 0, 19)).toBe('A19') // 좌상단
    expect(pointLabel(3, 3, 19)).toBe('D16') // 좌상 화점
    expect(pointLabel(3, 15, 19)).toBe('D4') // 좌하 화점
    expect(pointLabel(4, 4, 9)).toBe('E5') // 9줄 천원
  })

  it('skips the letter I', () => {
    expect(pointLabel(7, 0, 9)).toBe('H9')
    expect(pointLabel(8, 0, 9)).toBe('J9')
  })

  it('labels 19x19 star points the way Go books do', () => {
    const labels = starPoints(19).map((p) => pointLabel(p.x, p.y, 19))
    expect(labels).toContain('D16')
    expect(labels).toContain('Q16')
    expect(labels).toContain('K10') // 천원
    expect(labels).toContain('D4')
    expect(labels).toContain('Q4')
  })
})

describe('go engine', () => {
  it('places stones and alternates', () => {
    let g = createGame(9)
    const a = tryPlay(g, 3, 3)
    expect(a.ok).toBe(true)
    if (!a.ok) return
    g = a.state
    expect(g.toPlay).toBe(2)
    const b = tryPlay(g, 4, 3)
    expect(b.ok).toBe(true)
  })

  it('captures a single stone', () => {
    let g = createGame(9)
    // Black surrounds (1,1): north/west/east first, then white enters, black south captures
    const sequence: [number, number][] = [
      [0, 1], // B
      [5, 5], // W dummy
      [1, 0], // B
      [5, 6], // W dummy
      [2, 1], // B
      [1, 1], // W — the victim
      [1, 2], // B — capture
    ]
    let last = tryPlay(g, sequence[0][0], sequence[0][1])
    expect(last.ok).toBe(true)
    if (!last.ok) return
    g = last.state
    for (let i = 1; i < sequence.length; i++) {
      last = tryPlay(g, sequence[i][0], sequence[i][1])
      expect(last.ok).toBe(true)
      if (!last.ok) return
      g = last.state
    }
    expect(last.move.captured).toEqual([{ x: 1, y: 1 }])
    expect(g.captures[1]).toBe(1)
  })

  it('rejects suicide', () => {
    let g = createGame(9)
    // Black surrounds corner (0,0); White must not play there
    const setup: [number, number][] = [
      [1, 0], // B
      [5, 5], // W
      [0, 1], // B
      [5, 6], // W
    ]
    for (const [x, y] of setup) {
      const r = tryPlay(g, x, y)
      expect(r.ok).toBe(true)
      if (!r.ok) return
      g = r.state
    }
    // Black to play — pass so White faces suicide at (0,0)
    const p = pass(g)
    expect(p.ok).toBe(true)
    if (!p.ok) return
    g = p.state
    const suicide = tryPlay(g, 0, 0)
    expect(suicide.ok).toBe(false)
  })

  it('ends after two passes', () => {
    let g = createGame(9)
    let r = pass(g)
    if (!r.ok) throw new Error('pass')
    g = r.state
    r = pass(g)
    if (!r.ok) throw new Error('pass')
    expect(r.state.ended).toBe(true)
  })

  it('enforces simple ko', () => {
    let g = createGame(9)
    // Ko shape: White at (1,1); Black captures at (2,1) with single liberty → ko at (1,1)
    const sequence: [number, number][] = [
      [1, 0], // B
      [2, 0], // W
      [0, 1], // B
      [2, 2], // W
      [1, 2], // B
      [3, 1], // W
      [5, 5], // B dummy
      [1, 1], // W victim
      [2, 1], // B capture → ko
    ]
    for (const [x, y] of sequence) {
      const r = tryPlay(g, x, y)
      expect(r.ok).toBe(true)
      if (!r.ok) return
      g = r.state
    }
    expect(g.koPoint).toEqual({ x: 1, y: 1 })
    const illegal = tryPlay(g, 1, 1)
    expect(illegal.ok).toBe(false)
    if (illegal.ok) return
    expect(illegal.reason).toBe('ko')

    // Fill a move elsewhere to lift ko
    let r = tryPlay(g, 4, 4)
    expect(r.ok).toBe(true)
    if (!r.ok) return
    g = r.state
    r = tryPlay(g, 4, 5)
    expect(r.ok).toBe(true)
    if (!r.ok) return
    g = r.state
    r = tryPlay(g, 1, 1)
    // 패 직후가 아니면 슈퍼코/패 규칙에 따라 가능할 수 있음 — 여기선 한 수씩 둔 뒤 재시도
    expect(r.ok).toBe(true)
  })

  it('rejects positional superko (repeat board)', () => {
    let g = createGame(9)
    // 간단 순환: 따냄으로 이전 국면 재현 시도는 패/슈퍼코로 막힘
    const sequence: [number, number][] = [
      [1, 0],
      [2, 0],
      [0, 1],
      [2, 2],
      [1, 2],
      [3, 1],
      [5, 5],
      [1, 1],
      [2, 1], // B captures — ko at 1,1
    ]
    for (const [x, y] of sequence) {
      const r = tryPlay(g, x, y)
      expect(r.ok).toBe(true)
      if (!r.ok) return
      g = r.state
    }
    const retake = tryPlay(g, 1, 1)
    expect(retake.ok).toBe(false)
    if (retake.ok) return
    expect(['ko', 'superko']).toContain(retake.reason)
  })

  it('rejects suicide (fill own last liberty)', () => {
    let g = createGame(9)
    // White box around (1,1) with one hole; black cannot play into only liberty if suicide
    const sequence: [number, number][] = [
      [0, 0], // B elsewhere
      [1, 0], // W
      [5, 5], // B
      [0, 1], // W
      [5, 6], // B
      [2, 1], // W
      [5, 7], // B
      [1, 2], // W — (1,1) empty, surrounded by W on N/W/E/S
    ]
    for (const [x, y] of sequence) {
      const r = tryPlay(g, x, y)
      expect(r.ok).toBe(true)
      if (!r.ok) return
      g = r.state
    }
    // Black to play into (1,1) — suicide
    const suicide = tryPlay(g, 1, 1)
    expect(suicide.ok).toBe(false)
  })

  it('allows capturing instead of suicide', () => {
    let g = createGame(9)
    // Classic: white stone at (1,1) with one liberty; black captures
    const sequence: [number, number][] = [
      [1, 0], // B
      [1, 1], // W victim
      [0, 1], // B
      [5, 5], // W
      [2, 1], // B
      [5, 6], // W
    ]
    for (const [x, y] of sequence) {
      const r = tryPlay(g, x, y)
      expect(r.ok).toBe(true)
      if (!r.ok) return
      g = r.state
    }
    const cap = tryPlay(g, 1, 2) // B captures W
    expect(cap.ok).toBe(true)
    if (!cap.ok) return
    expect(cap.move.captured).toEqual([{ x: 1, y: 1 }])
  })
})

File: app/test/domain/board_test.dart
// test/domain/board_test.dart — src/engine/board.test.ts 이식 + 추가
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:svil_baduk/domain/engine/board.dart';
import 'package:svil_baduk/domain/engine/types.dart';
import 'package:svil_baduk/domain/engine/zobrist.dart';

/// 좌표 목록을 순서대로 둔다
GameState play(BoardSize size, List<(int, int)> moves) {
  GameState g = createGame(size);
  for (final (int x, int y) in moves) {
    final PlayResult r = tryPlay(g, x, y);
    if (r is! PlayOk) {
      fail('($x,$y) 착수 실패: ${(r as PlayErr).error}');
    }
    g = r.state;
  }
  return g;
}

void main() {
  group('기본 착수', () {
    test('교대로 둔다', () {
      final GameState g = play(BoardSize.s9, <(int, int)>[(3, 3)]);
      expect(g.toPlay, Stone.white);
      expect(g.stoneAt(3, 3), Stone.black);
      expect(tryPlay(g, 4, 3).ok, isTrue);
    });

    test('같은 자리에 두 번 두지 못한다', () {
      final GameState g = play(BoardSize.s9, <(int, int)>[(3, 3)]);
      final PlayResult r = tryPlay(g, 3, 3);
      expect(r, isA<PlayErr>());
      expect((r as PlayErr).error, MoveError.occupied);
    });

    test('판 밖은 거부', () {
      final GameState g = createGame(BoardSize.s9);
      expect((tryPlay(g, -1, 0) as PlayErr).error, MoveError.outOfBounds);
      expect((tryPlay(g, 9, 0) as PlayErr).error, MoveError.outOfBounds);
    });
  });

  group('따냄', () {
    test('한 점을 따낸다', () {
      // 흑이 (0,0) 백을 둘러싼다
      final GameState g = play(BoardSize.s9, <(int, int)>[
        (1, 0), // B
        (0, 0), // W
        (0, 1), // B — (0,0) 활로 0
      ]);
      expect(g.stoneAt(0, 0), Stone.empty);
      expect(g.blackCaptures, 1);
      expect(g.history.last.captured, <Point>[const Point(0, 0)]);
    });

    test('여러 점을 한꺼번에 따낸다', () {
      // 백 두 점 (0,0),(1,0) 을 흑이 둘러싼다
      GameState g = createGame(BoardSize.s9);
      for (final (int x, int y) in <(int, int)>[
        (2, 0), // B
        (0, 0), // W
        (0, 1), // B
        (1, 0), // W
        (5, 5), // B (여유수)
        (8, 8), // W (여유수)
      ]) {
        g = (tryPlay(g, x, y) as PlayOk).state;
      }
      final PlayResult r = tryPlay(g, 1, 1); // 흑이 마지막 활로를 막는다
      expect(r, isA<PlayOk>());
      final GameState after = (r as PlayOk).state;
      expect(after.stoneAt(0, 0), Stone.empty);
      expect(after.stoneAt(1, 0), Stone.empty);
      expect(after.blackCaptures, 2);
    });

    test('따내는 수는 자살수가 아니다', () {
      // 백 한 점이 흑에 둘러싸여 활로 1 — 흑이 그 활로에 두면 따낸다
      final GameState g = play(BoardSize.s9, <(int, int)>[
        (1, 0), // B
        (0, 0), // W
        (5, 5), // B
        (8, 8), // W
      ]);
      final PlayResult r = tryPlay(g, 0, 1);
      expect(r, isA<PlayOk>());
      expect((r as PlayOk).state.blackCaptures, 1);
    });
  });

  group('자살수', () {
    test('활로 없는 자리에 두지 못한다', () {
      // 백이 (0,0) 을 제외한 모서리를 감싼 상태에서 흑이 (0,0) 에 두면 자살
      GameState g = createGame(BoardSize.s9);
      for (final (int x, int y) in <(int, int)>[
        (5, 5), // B
        (1, 0), // W
        (6, 6), // B
        (0, 1), // W
      ]) {
        g = (tryPlay(g, x, y) as PlayOk).state;
      }
      // 이제 흑 차례, (0,0) 은 활로 0
      final PlayResult r = tryPlay(g, 0, 0);
      expect(r, isA<PlayErr>());
      expect((r as PlayErr).error, MoveError.suicide);
    });
  });

  group('패', () {
    test('되따냄을 즉시 금지한다', () {
      // 표준 패 모양을 배치로 만든다
      final GameState g = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[Point(3, 1), Point(2, 2), Point(4, 2), Point(3, 3)],
        white: const <Point>[Point(2, 1), Point(1, 2), Point(2, 3)],
        toPlay: Stone.white,
      );
      // 백이 (3,2) 에 두어 흑 (2,2) 를 따낸다
      final PlayResult take = tryPlay(g, 3, 2);
      expect(take, isA<PlayOk>());
      final GameState afterTake = (take as PlayOk).state;
      expect(afterTake.history.last.captured, <Point>[const Point(2, 2)]);
      expect(afterTake.koPoint, const Point(2, 2));

      // 흑이 바로 되따내려 하면 패로 금지
      final PlayResult back = tryPlay(afterTake, 2, 2);
      expect(back, isA<PlayErr>());
      expect((back as PlayErr).error, MoveError.ko);
    });

    test('다른 곳을 두면 패가 풀린다', () {
      final GameState g = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[Point(3, 1), Point(2, 2), Point(4, 2), Point(3, 3)],
        white: const <Point>[Point(2, 1), Point(1, 2), Point(2, 3)],
        toPlay: Stone.white,
      );
      GameState s = (tryPlay(g, 3, 2) as PlayOk).state;
      s = (tryPlay(s, 7, 7) as PlayOk).state; // 흑 팻감
      expect(s.koPoint, isNull);
    });
  });

  group('위치 초과 (superko)', () {
    test('배치 위치를 다시 만드는 수를 막는다', () {
      // 배치 해시가 시딩되지 않으면 이 수가 통과해 버린다 — 회귀 방지
      final GameState g = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[Point(3, 1), Point(2, 2), Point(4, 2), Point(3, 3)],
        white: const <Point>[Point(2, 1), Point(1, 2), Point(2, 3)],
        toPlay: Stone.white,
      );
      final GameState afterTake = (tryPlay(g, 3, 2) as PlayOk).state;
      // 패 금지를 우회해도 위치 초과로 걸려야 한다
      final GameState noKo = afterTake.copyWith(clearKo: true);
      final PlayResult back = tryPlay(noKo, 2, 2);
      expect(back, isA<PlayErr>());
      expect((back as PlayErr).error, MoveError.superko);
    });
  });

  group('패스와 종국', () {
    test('두 번 패스로 끝난다', () {
      GameState g = createGame(BoardSize.s9);
      g = (passMove(g) as PlayOk).state;
      expect(g.ended, isFalse);
      expect(g.consecutivePasses, 1);
      g = (passMove(g) as PlayOk).state;
      expect(g.ended, isTrue);
      expect(g.consecutivePasses, 2);
    });

    test('사이에 착수가 있으면 패스 카운트가 초기화된다', () {
      GameState g = createGame(BoardSize.s9);
      g = (passMove(g) as PlayOk).state;
      g = (tryPlay(g, 4, 4) as PlayOk).state;
      expect(g.consecutivePasses, 0);
      g = (passMove(g) as PlayOk).state;
      expect(g.ended, isFalse);
    });

    test('종국 후에는 둘 수 없다', () {
      GameState g = createGame(BoardSize.s9);
      g = (passMove(g) as PlayOk).state;
      g = (passMove(g) as PlayOk).state;
      expect((tryPlay(g, 0, 0) as PlayErr).error, MoveError.gameEnded);
    });

    test('기권하면 상대가 이긴다', () {
      final GameState g = resign(createGame(BoardSize.s9), Stone.black);
      expect(g.ended, isTrue);
      expect(g.resignedBy, Stone.black);
    });
  });

  group('좌표 표기', () {
    test('아래가 1이고 I 를 건너뛴다', () {
      expect(pointLabel(0, 18, 19), 'A1');
      expect(pointLabel(0, 0, 19), 'A19');
      // 19줄 좌상 화점 — 모든 바둑책이 D16 이라 부른다
      expect(pointLabel(3, 3, 19), 'D16');
      expect(pointLabel(3, 15, 19), 'D4');
      expect(pointLabel(7, 0, 9), 'H9');
      expect(pointLabel(8, 0, 9), 'J9');
    });

    test('화점이 표준 위치에 있다', () {
      final List<String> labels = starPoints(BoardSize.s19)
          .map((Point p) => pointLabel(p.x, p.y, 19))
          .toList();
      expect(labels, contains('D16'));
      expect(labels, contains('K10'));
      expect(labels, contains('Q4'));
    });
  });

  group('legalMoves', () {
    test('tryPlay 와 정확히 일치한다', () {
      final GameState g = play(BoardSize.s9, <(int, int)>[(4, 4), (3, 3), (5, 5)]);
      final Set<String> fast =
          legalMoves(g).map((Point p) => '${p.x},${p.y}').toSet();
      for (int y = 0; y < 9; y++) {
        for (int x = 0; x < 9; x++) {
          expect(fast.contains('$x,$y'), tryPlay(g, x, y).ok,
              reason: '($x,$y) 에서 불일치');
        }
      }
      expect(fast.length, 81 - 3);
    });
  });

  group('replayHistory', () {
    test('판·따냄·해시를 정확히 복원한다', () {
      final GameState g = play(BoardSize.s9, <(int, int)>[
        (1, 0), (0, 0), (0, 1), (5, 5), (8, 8), (2, 2),
      ]);
      final GameState replayed = replayHistory(BoardSize.s9, g.history);
      expect(replayed.board, g.board);
      expect(replayed.toPlay, g.toPlay);
      expect(replayed.blackCaptures, g.blackCaptures);
      expect(replayed.zobrist, g.zobrist);
    });
  });

  group('Zobrist', () {
    test('같은 배치는 같은 해시', () {
      final GameState a = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[Point(3, 3), Point(4, 4)],
        white: const <Point>[Point(5, 5)],
        toPlay: Stone.black,
      );
      final GameState b = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[Point(4, 4), Point(3, 3)], // 순서만 다름
        white: const <Point>[Point(5, 5)],
        toPlay: Stone.white, // 차례는 해시에 안 들어간다
      );
      expect(a.zobrist, b.zobrist);
    });

    test('다른 배치는 다른 해시', () {
      final GameState a = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[Point(3, 3)],
        white: const <Point>[],
        toPlay: Stone.black,
      );
      final GameState b = createProblemState(
        size: BoardSize.s9,
        black: const <Point>[],
        white: const <Point>[Point(3, 3)], // 같은 자리, 다른 색
        toPlay: Stone.black,
      );
      expect(a.zobrist, isNot(b.zobrist));
    });

    test('토글은 자기역원이다', () {
      const int h = 12345;
      final int once = Zobrist.toggle(h, 19, 3, 3, Stone.black);
      expect(Zobrist.toggle(once, 19, 3, 3, Stone.black), h);
    });

    test('시드가 고정이라 실행마다 같다', () {
      final Uint8ListLike board = Uint8ListLike(9);
      board.set(3, 3, Stone.black);
      final int h1 = Zobrist.ofBoard(board.data, 9);
      final int h2 = Zobrist.ofBoard(board.data, 9);
      expect(h1, h2);
      expect(h1, isNot(0));
    });
  });
}

/// 테스트 편의용 얇은 래퍼
class Uint8ListLike {
  Uint8ListLike(this.lines) : data = Uint8List(lines * lines);
  final int lines;
  final Uint8List data;
  void set(int x, int y, Stone s) => data[y * lines + x] = s.wire;
}
